#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "dialogcreervehicule.h"
#include "dialogcreercolis.h"
#include "dialogdetailsvehicule.h"
#include <QProgressBar>
#include <QHBoxLayout>
#include <QMessageBox>
#include <QTableWidgetItem>
#include <QBrush>
#include <QColor>
#include "../Drone.h"
#include "../CamionnetteElectrique.h"
#include "../RobotTrottoir.h"
#include "../ColisFragile.h"
#include "../ColisRefrigere.h"
#include "../ColisExpress.h"
#include "../ColisStandard.h"

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    
    // Configuration de la fen√™tre
    setWindowTitle("Gestion de Flotte de Livraison Autonome");
    resize(1200, 700);
    
    // Connexions des signaux/slots
    setupConnections();
    
    // Message de bienvenue
    QMessageBox::information(this, "Bienvenue", 
        "Bienvenue dans le syst√®me de gestion de flotte!\n\n"
        "Version Qt - Interface Graphique\n"
        "G√©rez vos drones, camionnettes et robots autonomes.");
    
    // Rafra√Æchir l'interface
    rafraichirVehicules();
    rafraichirColis();
    rafraichirStatistiques();
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::setupConnections()
{
    // Connexions du menu
    connect(ui->actionAPropos, &QAction::triggered, this, &MainWindow::onAPropos);
    
    // Connexions de la toolbar
    connect(ui->btnCreerVehicule, &QPushButton::clicked, this, &MainWindow::onCreerVehicule);
    connect(ui->btnCreerColis, &QPushButton::clicked, this, &MainWindow::onCreerColis);
    connect(ui->btnAssigner, &QPushButton::clicked, this, &MainWindow::onAssignerColis);
    connect(ui->btnLivrer, &QPushButton::clicked, this, &MainWindow::onLivrerVehicule);
    connect(ui->btnSupprimer, &QPushButton::clicked, this, &MainWindow::onSupprimerVehicule);
    connect(ui->btnDetails, &QPushButton::clicked, this, &MainWindow::onAfficherDetails);
}

void MainWindow::rafraichirVehicules()
{
    // D√©sactiver le tri pendant la mise √† jour
    ui->tableVehicules->setSortingEnabled(false);
    
    // Vider le tableau
    ui->tableVehicules->setRowCount(0);
    
    // Parcourir tous les v√©hicules de la flotte
    int nbVehicules = static_cast<int>(flotte_.getNombreVehicules());
    
    for (int i = 0; i < nbVehicules; ++i) {
        const Vehicule* v = flotte_.getVehicule(static_cast<size_t>(i));
        if (!v) continue;
        
        // Ajouter une nouvelle ligne
        int row = ui->tableVehicules->rowCount();
        ui->tableVehicules->insertRow(row);
        
        // D√©terminer le type de v√©hicule avec dynamic_cast
        QString type;
        QColor couleurType;
        
        if (dynamic_cast<const Drone*>(v)) {
            type = "Drone";
            couleurType = QColor("#2e7d32");  // Vert
        } else if (dynamic_cast<const CamionnetteElectrique*>(v)) {
            type = "Camionnette";
            couleurType = QColor("#0d47a1");  // Bleu
        } else if (dynamic_cast<const RobotTrottoir*>(v)) {
            type = "Robot";
            couleurType = QColor("#424242");  // Gris
        } else {
            type = "Inconnu";
            couleurType = QColor("#757575");
        }
        
        // Colonne 0: ID
        QTableWidgetItem* itemId = new QTableWidgetItem(QString::number(v->getId()));
        itemId->setTextAlignment(Qt::AlignCenter);
        ui->tableVehicules->setItem(row, 0, itemId);
        
        // Colonne 1: Nom
        QTableWidgetItem* itemNom = new QTableWidgetItem(QString::fromStdString(v->getNom()));
        ui->tableVehicules->setItem(row, 1, itemNom);
        
        // Colonne 2: Type (color√© selon le type)
        QTableWidgetItem* itemType = new QTableWidgetItem(type);
        itemType->setForeground(QBrush(couleurType));
        itemType->setTextAlignment(Qt::AlignCenter);
        ui->tableVehicules->setItem(row, 2, itemType);
        
        // Colonne 3: √âtat
        QString etatStr = QString::fromStdString(v->etatToString(v->getEtat()));
        QTableWidgetItem* itemEtat = new QTableWidgetItem(etatStr);
        itemEtat->setTextAlignment(Qt::AlignCenter);
        
        // Colorier selon l'√©tat
        if (v->getEtat() == Etat::EN_SERVICE) {
            itemEtat->setForeground(QBrush(QColor("#2e7d32")));  // Vert
        } else if (v->getEtat() == Etat::HORS_SERVICE) {
            itemEtat->setForeground(QBrush(QColor("#c62828")));  // Rouge
        } else {
            itemEtat->setForeground(QBrush(QColor("#0d47a1")));  // Bleu
        }
        
        ui->tableVehicules->setItem(row, 3, itemEtat);
        
        // Colonne 4: Capacit√© (format: X.XX / Y.YY kg)
        double capaciteUtilisee = v->getCapaciteUtilisee() / 1000.0;  // Conversion g -> kg
        double capaciteMax = v->getCapaciteMax() / 1000.0;
        QString capacite = QString("%1 / %2 kg")
                              .arg(capaciteUtilisee, 0, 'f', 2)
                              .arg(capaciteMax, 0, 'f', 2);
        
        QTableWidgetItem* itemCapacite = new QTableWidgetItem(capacite);
        itemCapacite->setTextAlignment(Qt::AlignCenter);
        
        // Tooltip d√©taill√©
        double capaciteDispo = capaciteMax - capaciteUtilisee;
        QString tooltip = QString("Capacit√© maximale: %1 kg\nUtilis√©e: %2 kg\nDisponible: %3 kg")
                             .arg(capaciteMax, 0, 'f', 2)
                             .arg(capaciteUtilisee, 0, 'f', 2)
                             .arg(capaciteDispo, 0, 'f', 2);
        itemCapacite->setToolTip(tooltip);
        
        ui->tableVehicules->setItem(row, 4, itemCapacite);
        
        // Colonne 5: Barre de progression de charge
        QWidget* widgetBarre = new QWidget();
        QHBoxLayout* layoutBarre = new QHBoxLayout(widgetBarre);
        layoutBarre->setContentsMargins(4, 2, 4, 2);
        
        QProgressBar* progressBar = new QProgressBar();
        int pourcentage = (capaciteMax > 0) ? (int)((capaciteUtilisee / capaciteMax) * 100) : 0;
        progressBar->setValue(pourcentage);
        progressBar->setFormat(QString("%1%").arg(pourcentage));
        progressBar->setTextVisible(true);
        progressBar->setMaximumHeight(20);
        
        // Couleur selon le taux de remplissage
        QString couleurBarre;
        if (pourcentage < 50) {
            couleurBarre = "#2e7d32";  // Vert
        } else if (pourcentage < 80) {
            couleurBarre = "#f57c00";  // Orange
        } else if (pourcentage < 95) {
            couleurBarre = "#ff6f00";  // Orange fonc√©
        } else {
            couleurBarre = "#c62828";  // Rouge
        }
        
        progressBar->setStyleSheet(
            QString("QProgressBar {"
                    "   border: 1px solid #3a3a3a;"
                    "   border-radius: 3px;"
                    "   text-align: center;"
                    "   background-color: #1e1e1e;"
                    "   color: #e0e0e0;"
                    "   font-size: 10px;"
                    "   font-weight: bold;"
                    "}"
                    "QProgressBar::chunk {"
                    "   background-color: %1;"
                    "   border-radius: 2px;"
                    "}").arg(couleurBarre)
        );
        
        progressBar->setToolTip(tooltip);
        layoutBarre->addWidget(progressBar);
        
        ui->tableVehicules->setCellWidget(row, 5, widgetBarre);
        
        // Colonne 6: Nombre de colis
        QTableWidgetItem* itemNbColis = new QTableWidgetItem(
            QString::number(v->getNombreColis())
        );
        itemNbColis->setTextAlignment(Qt::AlignCenter);
        itemNbColis->setToolTip(QString("%1 colis √† bord").arg(v->getNombreColis()));
        ui->tableVehicules->setItem(row, 6, itemNbColis);
    }
    
    // R√©activer le tri
    ui->tableVehicules->setSortingEnabled(true);
    
    // Ajuster les tailles de colonnes
    ui->tableVehicules->resizeColumnToContents(0);  // ID
    ui->tableVehicules->resizeColumnToContents(2);  // Type
    ui->tableVehicules->resizeColumnToContents(3);  // √âtat
    ui->tableVehicules->resizeColumnToContents(4);  // Capacit√©
    ui->tableVehicules->setColumnWidth(5, 120);     // Barre de progression
    ui->tableVehicules->resizeColumnToContents(6);  // Nb Colis
}

void MainWindow::rafraichirColis()
{
    // D√©sactiver le tri pendant la mise √† jour
    ui->tableColisAttente->setSortingEnabled(false);
    
    // Vider le tableau
    ui->tableColisAttente->setRowCount(0);
    
    // R√©cup√©rer le nombre de colis en attente
    size_t nbColis = flotte_.getNombreColisEnAttente();
    
    // Ajouter chaque colis
    for (size_t i = 0; i < nbColis; ++i) {
        const Colis* c = flotte_.getColis(i);
        if (!c) continue;
        
        int row = ui->tableColisAttente->rowCount();
        ui->tableColisAttente->insertRow(row);
        
        // D√©terminer le type de colis avec dynamic_cast
        QString type = "Standard";
        QString details = "-";
        
        if (auto* fragile = dynamic_cast<const ColisFragile*>(c)) {
            type = "Fragile";
            // Obtenir le niveau de fragilit√©
            switch (fragile->getFragilite()) {
                case Fragilite::FAIBLE:
                    details = "Fragilit√©: Faible";
                    break;
                case Fragilite::MOYEN:
                    details = "Fragilit√©: Moyen";
                    break;
                case Fragilite::ELEVE:
                    details = "Fragilit√©: √âlev√©";
                    break;
            }
        } else if (auto* refrigere = dynamic_cast<const ColisRefrigere*>(c)) {
            type = "R√©frig√©r√©";
            details = QString("Temp: %1¬∞C").arg(refrigere->getTemperatureCible(), 0, 'f', 1);
        } else if (dynamic_cast<const ColisExpress*>(c)) {
            type = "Express";
            details = "URGENT";
        }
        
        // Colonne 0: Index
        QTableWidgetItem* itemIndex = new QTableWidgetItem(QString::number(i));
        itemIndex->setTextAlignment(Qt::AlignCenter);
        ui->tableColisAttente->setItem(row, 0, itemIndex);
        
        // Colonne 1: Description
        QTableWidgetItem* itemDesc = new QTableWidgetItem(
            QString::fromStdString(c->getDescription())
        );
        QString tooltipDesc = QString("Description: %1\nMasse: %2 kg\nType: %3")
                                 .arg(QString::fromStdString(c->getDescription()))
                                 .arg(c->getMasse() / 1000.0, 0, 'f', 2)
                                 .arg(type);
        if (details != "-") {
            tooltipDesc += "\n" + details;
        }
        itemDesc->setToolTip(tooltipDesc);
        ui->tableColisAttente->setItem(row, 1, itemDesc);
        
        // Colonne 2: Masse (en kg)
        double masseKg = c->getMasse() / 1000.0;
        QTableWidgetItem* itemMasse = new QTableWidgetItem(
            QString("%1 kg").arg(masseKg, 0, 'f', 2)
        );
        itemMasse->setTextAlignment(Qt::AlignCenter);
        itemMasse->setToolTip(QString("Masse: %1 kg (%2 g)")
                                .arg(masseKg, 0, 'f', 2)
                                .arg(c->getMasse()));
        ui->tableColisAttente->setItem(row, 2, itemMasse);
        
        // Colonne 3: Type (colori√© selon le type)
        QTableWidgetItem* itemType = new QTableWidgetItem(type);
        itemType->setTextAlignment(Qt::AlignCenter);
        itemType->setToolTip(QString("Type de colis: %1").arg(type));
        
        // Colorier selon le type
        if (type == "Fragile") {
            itemType->setForeground(QBrush(QColor("#c62828")));  // Rouge
        } else if (type == "R√©frig√©r√©") {
            itemType->setForeground(QBrush(QColor("#0d47a1")));  // Bleu
        } else if (type == "Express") {
            itemType->setForeground(QBrush(QColor("#ff6f00")));  // Orange
        } else {
            itemType->setForeground(QBrush(QColor("#757575")));  // Gris
        }
        
        ui->tableColisAttente->setItem(row, 3, itemType);
        
        // Colonne 4: D√©tails
        QTableWidgetItem* itemDetails = new QTableWidgetItem(details);
        itemDetails->setTextAlignment(Qt::AlignCenter);
        if (details != "-") {
            itemDetails->setToolTip(details);
        }
        ui->tableColisAttente->setItem(row, 4, itemDetails);
    }
    
    // R√©activer le tri
    ui->tableColisAttente->setSortingEnabled(true);
    
    // Ajuster la largeur des colonnes
    ui->tableColisAttente->resizeColumnsToContents();
}

void MainWindow::rafraichirStatistiques()
{
    // Mise √† jour de la barre de statut
    int nbVehicules = flotte_.getNombreVehicules();
    int nbColis = flotte_.getNombreColisEnAttente();
    
    QString stats = QString("V√©hicules: %1 | Colis en attente: %2")
                        .arg(nbVehicules)
                        .arg(nbColis);
    
    statusBar()->showMessage(stats);
}

// ========== SLOTS DU MENU ==========

void MainWindow::onAPropos()
{
    QMessageBox::about(this, "√Ä propos",
        "<h2>Gestion de Flotte de Livraison Autonome</h2>"
        "<p>Version 1.0 - Interface Qt</p>"
        "<p>Syst√®me de gestion pour v√©hicules autonomes:</p>"
        "<ul>"
        "<li>üöÅ Drones (5 kg max)</li>"
        "<li>üöö Camionnettes √©lectriques (500 kg max)</li>"
        "<li>ü§ñ Robots de trottoir (20 kg max)</li>"
        "</ul>"
        "<p>D√©velopp√© avec C++17 et Qt Widgets</p>");
}

// ========== SLOTS DE LA TOOLBAR ==========

void MainWindow::onCreerVehicule()
{
    DialogCreerVehicule dialog(this);
    
    if (dialog.exec() == QDialog::Accepted) {
        // R√©cup√©rer le v√©hicule cr√©√©
        auto vehicule = dialog.getVehicule();
        
        if (vehicule) {
            // Ajouter √† la flotte
            flotte_.ajouterVehicule(std::move(vehicule));
            
            // Rafra√Æchir l'affichage
            rafraichirVehicules();
            rafraichirStatistiques();
            
            statusBar()->showMessage("V√©hicule cr√©√© avec succ√®s", 3000);
        }
    }
}

void MainWindow::onCreerColis()
{
    DialogCreerColis dialog(this);
    
    if (dialog.exec() == QDialog::Accepted) {
        // R√©cup√©rer le colis cr√©√©
        auto colis = dialog.getColis();
        
        if (colis) {
            // Ajouter √† la flotte (file d'attente)
            flotte_.recevoirColis(std::move(colis));
            
            // Rafra√Æchir l'affichage
            rafraichirColis();
            rafraichirStatistiques();
            
            statusBar()->showMessage("Colis ajout√© en attente d'assignation", 3000);
        }
    }
}

void MainWindow::onAssignerColis()
{
    // V√©rifier qu'un colis est s√©lectionn√©
    QList<QTableWidgetItem*> selectedColis = ui->tableColisAttente->selectedItems();
    if (selectedColis.isEmpty()) {
        QMessageBox::warning(this, "Aucune s√©lection", 
            "Veuillez s√©lectionner un colis dans la liste d'attente.");
        return;
    }
    
    // V√©rifier qu'un v√©hicule est s√©lectionn√©
    QList<QTableWidgetItem*> selectedVehicules = ui->tableVehicules->selectedItems();
    if (selectedVehicules.isEmpty()) {
        QMessageBox::warning(this, "Aucune s√©lection", 
            "Veuillez s√©lectionner un v√©hicule dans la flotte.");
        return;
    }
    
    // R√©cup√©rer l'index du colis (colonne 0)
    int rowColis = ui->tableColisAttente->currentRow();
    if (rowColis < 0) return;
    
    int indexColis = ui->tableColisAttente->item(rowColis, 0)->text().toInt();
    
    // R√©cup√©rer l'index du v√©hicule (ligne s√©lectionn√©e)
    int indexVehicule = ui->tableVehicules->currentRow();
    if (indexVehicule < 0) return;
    
    // Tenter l'assignation
    if (flotte_.assignerColis(indexColis, indexVehicule)) {
        // Succ√®s
        rafraichirVehicules();
        rafraichirColis();
        rafraichirStatistiques();
        
        statusBar()->showMessage("Colis assign√© avec succ√®s", 3000);
    } else {
        // √âchec
        QMessageBox::warning(this, "Assignation impossible", 
            "Le colis n'a pas pu √™tre assign√© au v√©hicule.\n"
            "V√©rifiez que le v√©hicule a suffisamment de capacit√© disponible.");
    }
}

void MainWindow::onLivrerVehicule()
{
    // V√©rifier qu'un v√©hicule est s√©lectionn√©
    QList<QTableWidgetItem*> selectedVehicules = ui->tableVehicules->selectedItems();
    if (selectedVehicules.isEmpty()) {
        QMessageBox::warning(this, "Aucune s√©lection", 
            "Veuillez s√©lectionner un v√©hicule dans la flotte.");
        return;
    }
    
    // R√©cup√©rer l'index du v√©hicule
    int indexVehicule = ui->tableVehicules->currentRow();
    if (indexVehicule < 0) return;
    
    // V√©rifier que le v√©hicule a des colis
    const Vehicule* v = flotte_.getVehicule(indexVehicule);
    if (!v) return;
    
    if (v->getNombreColis() == 0) {
        QMessageBox::information(this, "Aucun colis", 
            "Ce v√©hicule ne transporte aucun colis.");
        return;
    }
    
    // Demander confirmation
    QString nomVehicule = QString::fromStdString(v->getNom());
    int nbColis = v->getNombreColis();
    
    QMessageBox::StandardButton reponse = QMessageBox::question(this, 
        "Confirmer la livraison",
        QString("Voulez-vous livrer les %1 colis du v√©hicule '%2' ?").arg(nbColis).arg(nomVehicule),
        QMessageBox::Yes | QMessageBox::No);
    
    if (reponse == QMessageBox::Yes) {
        // Effectuer la livraison
        flotte_.livrerVehicule(indexVehicule);
        
        // Rafra√Æchir
        rafraichirVehicules();
        rafraichirStatistiques();
        
        statusBar()->showMessage(
            QString("Livraison effectu√©e : %1 colis livr√©s").arg(nbColis), 
            3000
        );
    }
}

void MainWindow::onSupprimerVehicule()
{
    // V√©rifier qu'un v√©hicule est s√©lectionn√©
    QList<QTableWidgetItem*> selectedVehicules = ui->tableVehicules->selectedItems();
    if (selectedVehicules.isEmpty()) {
        QMessageBox::warning(this, "Aucune s√©lection", 
            "Veuillez s√©lectionner un v√©hicule dans la flotte.");
        return;
    }
    
    // R√©cup√©rer l'index du v√©hicule
    int indexVehicule = ui->tableVehicules->currentRow();
    if (indexVehicule < 0) return;
    
    // R√©cup√©rer les infos du v√©hicule
    const Vehicule* v = flotte_.getVehicule(indexVehicule);
    if (!v) return;
    
    QString nomVehicule = QString::fromStdString(v->getNom());
    int nbColis = v->getNombreColis();
    
    // V√©rifier si le v√©hicule a des colis
    if (nbColis > 0) {
        QMessageBox::warning(this, "Suppression impossible", 
            QString("Le v√©hicule '%1' transporte encore %2 colis.\n"
                   "Livrez ou retirez les colis avant de supprimer le v√©hicule.")
                   .arg(nomVehicule).arg(nbColis));
        return;
    }
    
    // Demander confirmation
    QMessageBox::StandardButton reponse = QMessageBox::question(this, 
        "Confirmer la suppression",
        QString("Voulez-vous vraiment supprimer le v√©hicule '%1' ?").arg(nomVehicule),
        QMessageBox::Yes | QMessageBox::No);
    
    if (reponse == QMessageBox::Yes) {
        // Supprimer le v√©hicule
        if (flotte_.supprimerVehicule(indexVehicule)) {
            // Rafra√Æchir
            rafraichirVehicules();
            rafraichirStatistiques();
            
            statusBar()->showMessage(
                QString("V√©hicule '%1' supprim√© de la flotte").arg(nomVehicule), 
                3000
            );
        }
    }
}

void MainWindow::onAfficherDetails()
{
    // V√©rifier qu'un v√©hicule est s√©lectionn√©
    QList<QTableWidgetItem*> selectedVehicules = ui->tableVehicules->selectedItems();
    if (selectedVehicules.isEmpty()) {
        QMessageBox::warning(this, "Aucune s√©lection", 
            "Veuillez s√©lectionner un v√©hicule dans la flotte.");
        return;
    }
    
    // R√©cup√©rer l'index du v√©hicule
    int indexVehicule = ui->tableVehicules->currentRow();
    if (indexVehicule < 0) return;
    
    // R√©cup√©rer le v√©hicule
    const Vehicule* v = flotte_.getVehicule(indexVehicule);
    if (!v) return;
    
    // Ouvrir le dialogue des d√©tails
    DialogDetailsVehicule dialogue(v, this);
    dialogue.exec();
}
